SparkPost/2web2ui
# FE-1154 - [Domains UX Redesign] List page - Make tables filterable

[FE-1154](https://sparkpost.atlassian.net/browse/FE-1154)

### What Changed

- Adds filtering behavior for the Sending, Bounce, and Tracking domains table for the Domains feature re-design
- Adds new, re-usable hook `useTable` hook that helps with the creation of composable table and table-like interfaces:
  + Adds supporting tests
  + Adds Storybook documentation
- Adds new `Table.SortButton` component for use with `useTable`
- Adds new index file to `src/hooks` so that hooks can be imported like so:
```javascript
import { useTable } from 'src/hooks'
```
- Adds new, re-usable component `Divider`
- Fixes flaky reporter builder Cypress test that was failing the build - updates a component in reporter builder to use correct underline HTML semantics from an a11y perspective

### How To Test

- Go to `/domains/list/sending`, `/domains/list/bounce`, and `/domains/list/tracking` and verify the tables filter as expected
- Do you like the API for `useTable`? if you don't, let's talk about it!

### To Do

- [ ] Cut ticket to Introduce debounce in to table filter
- [ ] Cut ticket to add pagination support to `useTable`
- [ ] Cut ticket to refactor fugly `filterByCollectionValues` array helper
- [ ] Replace "Failed Verification" tag with "Unverified"
- [x] Fix tracking domains table implementation now that `useTable` is available
- [x] Write automated tests around `useTable`
- [x] Date not stable locally vs. in CI
- [x] Fix double border in domain status popover
- [x] Content between buttons needs more space
- [x] Remove "Select All" checkbox
- [x] When no subaccount is on a row, render "Primary account"
- [x] When no date is present render "Creation date: unknown"
- [x] Reduce font size for subaccount and creation date using design token provided by UX: **fontSize.200, lineHeight.200**
- [x] Add unit tests for updated selectors
- [x] Some funky state problems when rapidly switching between tabs - disable while `pending`?
- [x] Handle table sorting via `useTable`
- [x] Handle table filtering via `useTable`
- [x] Add button content population for checkbox filters
- [x] Add unit tests for `Table.SortButton` and `Table.SortIcon`
- [x] Write docs for `useTable` with examples
- [x] Refactor `SendingDomainsTable` to reduce its length - probably have a `SendingDomainsTab` component as well
- [x] Fix sorting such that initially selected filters apply
- [x] The layout on smaller screen sizes is a little rough
- [x] Fix rendering "subaccount default" bounce domains
- [x] Fix simultaneous filtering and sorting
- [x] Incorporate `<fieldset>` and `<legend>` for the domain status checkboxes
- [x] Cypress tests for:
  + [x] Sending domains filtering/sorting
  + [x] Bounce domains filtering/sorting
  + [x] Tracking domains filtering/sorting
